core/drivers/gic.c

```diff
@@ -390,33 +402,59 @@ static void gic_it_set_pending(struct gic_data *gd, size_t it)
 }
 
 static void gic_it_raise_sgi(struct gic_data *gd __maybe_unused, size_t it,
-			     uint8_t cpu_mask, uint8_t group)
+			     uint32_t cpu_mask, uint8_t group)
 {
 #if defined(CFG_ARM_GICV3)
-	/* Only support sending SGI to the cores in the same cluster now */
 	uint32_t mask_id = it & 0xf;
-	uint32_t mask_cpu = cpu_mask & 0xff;
-	uint64_t mpidr = read_mpidr();
-	uint64_t mask_aff1 = (mpidr & MPIDR_AFF1_MASK) >> MPIDR_AFF1_SHIFT;
-	uint64_t mask_aff2 = (mpidr & MPIDR_AFF2_MASK) >> MPIDR_AFF2_SHIFT;
-	uint64_t mask_aff3 = (mpidr & MPIDR_AFF3_MASK) >> MPIDR_AFF3_SHIFT;
-	uint64_t mask = (mask_cpu |
-			SHIFT_U64(mask_aff1, 16) |
-			SHIFT_U64(mask_id, 24)   |
-			SHIFT_U64(mask_aff2, 32) |
-			SHIFT_U64(mask_aff3, 48));
+	uint64_t mask = SHIFT_U64(mask_id, 24);
+
```

Comment:
That expression is quite hard to parse and understand. How about:
``` C
bool to_this_cpu = cpu_mask & ITR_CPU_MASK_TO_THIS_CPU;
bool to_other_cpus = cpu_mask & ITR_CPU_MASK_TO_OTHER_CPUS;

assert(cpu_mask && (to_this_cpu + to_other_cpus) < 2);
```
The helper variables can then be used below too.

Reply:
Using "assert(cpu_mask && (to_this_cpu + to_other_cpus) < 2);" ,maybe missing a check if another bit  is set along with one of those? ITR_CPU_MASK_TO_THIS_CPU and ITR_CPU_MASK_TO_OTHER_CPUS should be exclusive with the rest of the bitfield, right?

Reply:
It depends on how far you want to take this. Having a monster of an assert expression will not make anyone happy. It must be readable.

Reply:
My 2 cents: when the assert expression is over complex, it can be handy to implement it in a helper function with an explicit name that tell what's the monster aims to.

Reply:
Would you be ok with:

`static bool gic_flags_are_exclusive(uint32_t cpu_mask)
{
	if (((cpu_mask &
	      (ITR_CPU_MASK_TO_THIS_CPU | ITR_CPU_MASK_TO_OTHER_CPUS)) &&
	     (!!(cpu_mask & ITR_CPU_MASK_TO_THIS_CPU) ^
	      !!(cpu_mask & ITR_CPU_MASK_TO_OTHER_CPUS)) &&
	     !(cpu_mask &
	       ~(ITR_CPU_MASK_TO_THIS_CPU | ITR_CPU_MASK_TO_OTHER_CPUS))) ||
	    (!(cpu_mask &
	       (ITR_CPU_MASK_TO_THIS_CPU | ITR_CPU_MASK_TO_OTHER_CPUS))))
		return true;

	return false;

...

	/* Assert exclusivity of mask hint bits and effective CPU mask bits */
	assert(cpu_mask && gic_flags_are_exclusive(cpu_mask));
}`

Reply:
I prefer Jens' proposal. Extended to check hint bits and cpu list bitfield are exclusive: 
```c
static bool assert_cpu_mask_is_valid(uint32_t cpu_mask)
{
	bool __maybe_unused to_others = cpu_mask & ITR_CPU_MASK_TO_OTHER_CPUS;
	bool __maybe_unused to_current = cpu_mask & ITR_CPU_MASK_TO_THIS_CPU;
	bool __maybe_unused to_list = cpu_mask & 0xff;

	assert(to_others + to_current + to_list == 1);
}

Reply:
Fine with a helper function, but the expression in that function is digested easily.

Reply:
That assert just begs for a _"There can be only one."_ comment :-)

Reply:
Ok for the last proposal + the comment then :)